lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -48,82 +47,38 @@ namespace scftocalyx {
 /// Calyx component.
 using FuncMapping = DenseMap<FuncOp, calyx::ComponentOp>;
 
-/// A structure that abstracts a WhileOp or PipelineWhileOp to provide a
-/// consistent interface.
-struct WhileOpInterface {
-  WhileOpInterface(Operation *op) {
-    assert(isSupported(op));
-    if (auto scfWhile = dyn_cast<scf::WhileOp>(op))
-      impl = scfWhile;
-    if (auto staticLogicWhile = dyn_cast<staticlogic::PipelineWhileOp>(op))
-      impl = staticLogicWhile;
-  }
-
-  static bool isSupported(Operation *op) {
-    return isa<scf::WhileOp, staticlogic::PipelineWhileOp>(op);
-  }
+class ScfWhileOp : public calyx::WhileOpInterface<scf::WhileOp> {
+public:
+  explicit ScfWhileOp(Operation *op)
+      : calyx::WhileOpInterface<scf::WhileOp>(op) {}
 
   Block::BlockArgListType getBodyArgs() {
-    return TypeSwitch<Operation *, Block::BlockArgListType>(impl)
-        .Case([](scf::WhileOp op) { return op.getAfterArguments(); })
-        .Case([](staticlogic::PipelineWhileOp op) {
-          return op.getStagesBlock().getArguments();
-        });
-    ;
+    return getOperation().getAfterArguments();
   }
 
-  Block *getBodyBlock() {
-    return TypeSwitch<Operation *, Block *>(impl)
-        .Case([](scf::WhileOp op) { return &op.getAfter().front(); })
-        .Case([](staticlogic::PipelineWhileOp op) {
-          return &op.getStagesBlock();
-        });
-  }
+  Block *getBodyBlock() { return &getOperation().getAfter().front(); }
```

Comment:
I'd expect an `override` specifier for these, but i guess if `clang-tidy` is not complaining, then this form is acceptable style.

Reply:
Good call. I've fixed this.